extensions/sftp/client/SFTPClient.cpp

```diff
@@ -0,0 +1,713 @@
+  LIBSSH2_SFTP_HANDLE *file_handle = libssh2_sftp_open(sftp_session_, path.c_str(), flags, 0644);
+  if (file_handle == nullptr) {
+    logger_->log_error("Failed to open remote file \"%s\", error: %s", path.c_str(), sftp_strerror(libssh2_sftp_last_error(sftp_session_)));
+    return false;
+  }
+  ScopeGuard guard([this, &file_handle, &path]() {
+    logger_->log_trace("Closing remote file \"%s\"", path.c_str());
+    libssh2_sftp_close(file_handle);
+  });
+
+  /* If they just want a zero byte file, we are done */
+  if (expected_size == 0) {
+    return true;
+  }
+
+  const size_t buf_size = expected_size < 0 ? MAX_BUFFER_SIZE : std::min<size_t>(expected_size, MAX_BUFFER_SIZE);
+  std::vector<uint8_t> buf(buf_size);
+  uint64_t total_read = 0U;
+  do {
+    int read_ret = input.readData(buf.data(), buf.size());
+    if (read_ret < 0) {
+      char *err_msg = nullptr;
+      libssh2_session_last_error(ssh_session_, &err_msg, nullptr, 0);
+      logger_->log_error("Error while reading input");
+      return false;
+    } else if (read_ret == 0) {
+      logger_->log_trace("EOF while reading input");
+      break;
+    }
+    logger_->log_trace("Read %d bytes", read_ret);
+    total_read += read_ret;
+    ssize_t remaining = read_ret;
+    while (remaining > 0) {
+      int write_ret = libssh2_sftp_write(file_handle, reinterpret_cast<char*>(buf.data() + (read_ret - remaining)), remaining);
+      if (write_ret < 0) {
+        logger_->log_error("Failed to write remote file \"%s\", error: %s", path.c_str(), sftp_strerror(libssh2_sftp_last_error(sftp_session_)));
+        return false;
+      }
+      logger_->log_trace("Wrote %d bytes to remote file \"%s\"", write_ret, path.c_str());
+      remaining -= write_ret;
+    }
+  } while (true);
+
+  if (expected_size >= 0 && total_read != expected_size) {
+    logger_->log_error("Input has unexpected size, expected: %ld, actual: %lu", path.c_str(), expected_size, total_read);
+    return false;
+  }
+
+  return true;
+}
+
+bool SFTPClient::rename(const std::string& source_path, const std::string& target_path, bool overwrite) {
+  int flags = 0;
+  if (overwrite) {
+    flags = LIBSSH2_SFTP_RENAME_ATOMIC | LIBSSH2_SFTP_RENAME_NATIVE | LIBSSH2_SFTP_RENAME_OVERWRITE;
+  } else {
+    flags = LIBSSH2_SFTP_RENAME_ATOMIC | LIBSSH2_SFTP_RENAME_NATIVE;
+  }
+  bool tried_deleting = false;
+  while (libssh2_sftp_rename_ex(sftp_session_,
+                              source_path.c_str(),
+                              source_path.length(),
+                              target_path.c_str(),
+                              target_path.length(),
+                              flags) != 0) {
+    auto err = libssh2_sftp_last_error(sftp_session_);
+    if (overwrite && err == LIBSSH2_FX_FILE_ALREADY_EXISTS && !tried_deleting) {
+      /* It couldn't overwrite the file, let's delete it and try again */
+      tried_deleting = true;
+      if (!this->removeFile(target_path)) {
+        return false;
```

Comment:
I agree, but I would only do it on a debug level, not error, because in my experience if a system does not support overwriting files with rename, it will consistently refuse it, and the logs would be full of this not-really-error.

Reply:
Added with log_debug.